### PR TITLE
Cohere Client

### PR DESCRIPTION
**Pull request by Hk669**
## Why are these changes needed?

To enhance the support of non-OpenAI models with AutoGen. The Command family of models includes [Command](https://cohere.com/models/command), [Command R](https://docs.cohere.com/docs/command-r), and [Command R+](https://docs.cohere.com/docs/command-r-plus). Together, they are the text-generation LLMs powering conversational agents, summarization, copywriting, and similar use cases.

- [x] Pricing Implemented
- [x] API Checks
- [x] Create
- [x] tool calls
- [x] get_usage

## Checks

- [x] I've included any doc changes needed for https://microsoft.github.io/autogen/. See https://microsoft.github.io/autogen/docs/Contribute#documentation to build and test documentation locally.
- [x] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [x] I've made sure all auto checks have passed.

### CODE REVIEW

File: autogen/oai/cohere.py [hunk 1]
```diff
@@ -0,0 +1,427 @@
...
+        max_retries = 5
+        for attempt in range(max_retries):
+            ans = None
+            try:
+                if streaming:
+                    response = client.chat_stream(**cohere_params)
+                else:
+                    response = client.chat(**cohere_params)
+            except Exception as e:
+                raise RuntimeError(f"Cohere exception occurred: {e}")
```
Comment (Hk669) [inline autogen/oai/cohere.py:170]: i dont think this is required. would love some feedback from @yiranwu0 @qingyun-wu

File: autogen/oai/cohere.py [hunk 2]
```diff
@@ -0,0 +1,427 @@
...
+            else:
+
+                if streaming:
+                    # Streaming...
+                    ans = ""
+                    for event in response:
+                        if event.event_type == "text-generation":
+                            ans = ans + event.text
```
Comment (Hk669) [inline autogen/oai/cohere.py:178]: imo we can follow the same method with other clients too, to enable the streaming option.

**Comment by marklysze** [comment]
Added Cohere documentation. We're missing Groq, so we will add that to here as well.

**Comment by marklysze** [comment]
@qingyun-wu, I've added the Groq documentation into this PR. I think we're good for a review now that we have documentation and tests for Cohere.

**Review by qingyun-wu — APPROVED** [review]
Looks great! Thanks!

File: autogen/oai/cohere.py [hunk 3]
```diff
@@ -0,0 +1,432 @@
...
+        cohere_params["p"] = validate_parameter(params, "p", (int, float), False, 0.75, (0.01, 0.99), None)
+        cohere_params["seed"] = validate_parameter(params, "seed", int, True, None, None, None)
+        cohere_params["frequency_penalty"] = validate_parameter(
+            params, "frequency_penalty", (int, float), True, 0, (0, 1), None
+        )
+        cohere_params["presence_penalty"] = validate_parameter(
+            params, "presence_penalty", (int, float), True, 0, (0, 1), None
+        )
+
+        # Cohere parameters we are ignoring:
```
Comment (qingyun-wu) [inline autogen/oai/cohere.py:116]: Could we log these ignored parameters?

File: autogen/oai/cohere.py (hunk 1 above)
Comment (qingyun-wu) [reply autogen/oai/cohere.py:164]: @Hk669, which part you think is not required?

**Comment by qingyun-wu** [comment]
@yiranwu0 could you help review?

File: autogen/oai/cohere.py (hunk 1 above)
Comment (Hk669) [reply autogen/oai/cohere.py:164]: the `max_retries` loop, i dont think we follow the same for the other clients. let it throw an exception. 

cc @marklysze

File: autogen/oai/cohere.py (hunk 3 above)
Comment (Hk669) [reply autogen/oai/cohere.py:116]: sure. let me add them to the log info.

File: autogen/oai/cohere.py (hunk 1 above)
Comment (marklysze) [reply autogen/oai/cohere.py:179]: Gemini has a [retries loop](https://github.com/microsoft/autogen/blob/8133b7de22a7b9a7c84cee31bbd963798efef753/autogen/oai/gemini.py#L193) but not all do. If we want to standardise on no loop, that's okay with me.

File: autogen/oai/cohere.py [hunk 4]
```diff
@@ -0,0 +1,447 @@
...
+
+        max_retries = 5
+        for attempt in range(max_retries):
+            ans = None
+            try:
+                if streaming:
+                    response = client.chat_stream(**cohere_params)
+                else:
+                    response = client.chat(**cohere_params)
+            except Exception as e:
```
Comment (sonichi) [inline autogen/oai/cohere.py:178]: Can we use a more specific exception class? We typically avoid using the generic Exception class.

Comment (Hk669) [reply autogen/oai/cohere.py]: updated with a Custom Exception `CohereRateLimitError`. thanks for reviewing @sonichi

File: autogen/oai/cohere.py (hunk 1 above)
Comment (Hk669) [reply autogen/oai/cohere.py]: for now, let's maintain the current approach and adhere to the client's SDK as specified. thank you @marklysze

**Comment by Anirudh31415926535** [comment]
Hi, the Cohere integration looks fine on this PR! Thanks so much for the contribution! :)

**Pull request merged**